### PR TITLE
Only send verifications requests to devices that are cross-signed

### PR DESCRIPTION
**Pull request by zecakeh**
The commit adds a method that allows to check first if any device is cross-signed.

cc @poljar.

Signed-off-by: Kévin Commaille <zecakeh@tedomum.fr>

### CODE REVIEW

**Comment by zecakeh** [comment]
I tested this locally on Fractal with two accounts: one that has another device that is not cross-signed, and one that other devices that are cross-signed and in both cases I got the expected outcome.

**Comment by zecakeh** [comment]
I tested this locally on Fractal with two accounts: one that has another device that is not cross-signed, and one that has other devices that are cross-signed and in both cases I got the expected outcome.

File: crates/matrix-sdk-crypto/src/identities/user.rs [hunk 1]
```diff
@@ -166,8 +166,12 @@ impl OwnUserIdentity {
             .store
             .get_user_devices(self.user_id())
             .await?
-            .into_keys()
-            .filter(|d| &**d != self.verification_machine.own_device_id())
+            .into_iter()
+            .filter_map(|(d_id, d)| {
+                (d_id != self.verification_machine.own_device_id()
```
Comment (poljar) [inline crates/matrix-sdk-crypto/src/identities/user.rs:171]: Let's move this into a separate method, that way this also becomes testable. I don't require a test but would be nice.

**Review by poljar — CHANGES_REQUESTED** [review]

File: crates/matrix-sdk-crypto/src/identities/user.rs (hunk 1 above)
Comment (zecakeh) [reply crates/matrix-sdk-crypto/src/identities/user.rs:171]: What would the signature look like?

1. `fn should_send_request_to_device(&self, device_id: &DeviceId, own_device_id: &DeviceId, device: &ReadOnlyDevice) -> bool`
2. `fn devices_to_request(&self, devices: HashMap<OwnedDeviceId, ReadOnlyDevice>, own_device_id: &DeviceId) -> Vec<OwnedDeviceId>`

Or something else even?

Comment (poljar) [reply crates/matrix-sdk-crypto/src/identities/user.rs:171]: I think `UserDevices::signed_devices(&self) -> Iterator<Item = Device>` would be the most pleasant way to go about this.

Comment (zecakeh) [reply crates/matrix-sdk-crypto/src/identities/user.rs:171]: We don't have access to `UserDevices` here. It's a type returned by the `Store`, we only have access to the `VerificationStore` that wraps a `dyn CryptoStore` that returns a `HashMap<OwnedDeviceId, ReadOnlyDevice>`, afaict.

Comment (poljar) [reply crates/matrix-sdk-crypto/src/identities/user.rs:171]: Ah right, that's a shame. I guess option 2. from your proposal then sounds fine.

File: crates/matrix-sdk-crypto/src/identities/user.rs [hunk 2]
```diff
@@ -622,6 +619,19 @@ impl ReadOnlyOwnUserIdentity {
 
         Ok(())
     }
+
+    fn filter_devices_to_request(
+        &self,
+        devices: HashMap<OwnedDeviceId, ReadOnlyDevice>,
```
Comment (Hywan) [inline crates/matrix-sdk-crypto/src/identities/user.rs:625]: Is it mandatory to takes `devices` by value here? Could a reference work?

File: crates/matrix-sdk-crypto/src/identities/user.rs [hunk 3]
```diff
@@ -622,6 +619,19 @@ impl ReadOnlyOwnUserIdentity {
...
+        own_device_id: &DeviceId,
+    ) -> Vec<OwnedDeviceId> {
+        devices
+            .into_iter()
+            .filter_map(|(d_id, d)| {
+                (d_id != own_device_id && self.is_device_signed(&d).is_ok()).then_some(d_id)
+            })
```
Comment (Hywan) [inline crates/matrix-sdk-crypto/src/identities/user.rs:632]:
```suggestion
            .filter_map(|(device_id, device)| {
                (device_id != own_device_id && self.is_device_signed(&device).is_ok()).then_some(device_id)
            })
```

File: crates/matrix-sdk-crypto/src/identities/user.rs (hunk 2 above)
Comment (zecakeh) [reply crates/matrix-sdk-crypto/src/identities/user.rs:625]: We need to return a `Vec<OwnedDeviceId>` so this avoids to clone the IDs, especially given that the list is unused after.

Comment (Hywan) [reply crates/matrix-sdk-crypto/src/identities/user.rs:625]: Good to me then.

**Review by Hywan — APPROVED** [review]

**Review by poljar — APPROVED** [review]

**Pull request merged**